### PR TITLE
Replaced `hyper` + `tokio` with our own custom API server implementation

### PR DESCRIPTION
**Pull request by georgepisaltu**
## Reason for This PR

Fixes: #1193 #1194 #615 

## Description of Changes

The old `hyper` implementation of the API server is now replaced with another which uses our own `HttpServer`. The old `api_server` interface is refactored. It is now much easier to add API changes to Firecracker, as demonstrated in the commits.

Upsides
- release build time down from 2m09s to 1m49s (-21.6%)
- debug build time down from 50s to 40s (-20%)
- release binary size down from 2.956 KiB to 2.659 KiB (-10.1%)
- debug binary size down from 47.4 MiB to 36.5 MiB (-23%)
- resident memory usage on a basic configuration down from 1952 KiB to 1468 KiB (-24.8%)
- 956 fewer lines of code (+1091-55 LOC in this PR are the `HttpServer` implementation)
- 30 fewer dependencies, down from 104 to 74 (-28.8%)
- average request response time down from 1.122 ms/req to 0.766 ms/req (-31.7%)
- no more `future`s :sparkles:

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist
- [x] All commits in this PR are signed (`git commit -s`).
- [x] This PR is linked to multiple issues. 
- [x] The description of changes is clear and encompassing.
- [x] No docs need to be updated as part of this PR.
- [x] Code-level documentation for touched code is included in this PR.
- [x] No API changes are included in this PR`. :satisfied:
- [x] The changes in this PR have no user impact.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -14,22 +14,19 @@ version = "0.1.0"
...
  "logger 0.1.0",
  "memory_model 0.1.0",
+ "micro_http 0.1.0",
  "mmds 0.1.0",
  "net_util 0.1.0",
  "rate_limiter 0.1.0",
- "serde 1.0.101 (registry+https://github.com/rust-lang/crates.io-index)",
- "serde_derive 1.0.101 (registry+https://github.com/rust-lang/crates.io-index)",
- "serde_json 1.0.41 (registry+https://github.com/rust-lang/crates.io-index)",
+ "serde 1.0.99 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (acatangiu) [inline Cargo.lock]: During rebase you reverted the `cargo update` from the v0.19.0 release PR.
You need to run a `cargo update` here.

File: api_server/src/parsed_request.rs [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+pub enum ParsedRequest {
+    GetInstanceInfo,
+    GetMMDS,
+    PatchMMDS(Value),
+    PutMMDS(Value),
+    Sync(VmmAction),
+}
+
+#[cfg(test)]
+impl PartialEq for ParsedRequest {
```
Comment (acatangiu) [inline api_server/src/parsed_request.rs]: Move this to `mod tests`.

File: api_server/src/lib.rs [hunk 3]
```diff
@@ -157,43 +121,48 @@ impl fmt::Debug for Error {
...
     vmm_shared_info: Arc<RwLock<InstanceInfo>>,
-    // Sender which allows passing messages to the VMM.
-    api_request_sender: Rc<mpsc::Sender<VmmRequest>>,
-    efd: Rc<EventFd>,
+    /// Sender which allows passing messages to the VMM.
+    api_request_sender: mpsc::Sender<VmmRequest>,
+    /// Receiver which collects messages from the VMM.
+    vmm_response_receiver: mpsc::Receiver<VmmResponse>,
+    /// FD on which we notify the VMM that we have sent at least one
+    /// `VmmActionRequest`.
```
Comment (acatangiu) [inline api_server/src/lib.rs]:
```suggestion
    /// `VmmRequest`.
```

File: api_server/src/request/drive.rs [hunk 4]
```diff
@@ -1,16 +1,17 @@
...
-use vmm::vmm_config::drive::BlockDeviceConfig;
+use super::super::VmmAction;
+use logger::{Metric, METRICS};
+use request::Body;
+use request::Error;
+use request::StatusCode;
 
-use request::{IntoParsedRequest, ParsedRequest};
+use request::checked_id;
+use request::ParsedRequest;
```
Comment (acatangiu) [inline api_server/src/request/drive.rs]: nit: collapse these multiple lines to:
```
use request::{Body, Error, etc.};
```

File: api_server/src/request/drive.rs [hunk 5]
```diff
@@ -74,222 +81,213 @@ impl PatchDrivePayload {
...
-                Ok(ParsedRequest::Sync(
-                    VmmRequest::new(
-                        VmmAction::UpdateBlockDevicePath(drive_id, path_on_host),
-                        sender,
-                    ),
-                    receiver,
-                ))
-            }
-            _ => Err(format!("Invalid method {}!", method)),
+    if let Some(body) = maybe_body {
```
Comment (acatangiu) [inline api_server/src/request/drive.rs]: Most requests do this check against `Option<&Body>` and all return the same error on `None`.

Maybe add a static function to de-duplicate this code.

File: api_server/src/request/drive.rs [hunk 6]
```diff
@@ -74,222 +81,213 @@ impl PatchDrivePayload {
...
         }
+
+        Ok(ParsedRequest::Sync(VmmAction::UpdateBlockDevicePath(
+            drive_id,
+            path_on_host,
+        )))
+    } else {
+        Err(Error::Generic(
+            StatusCode::BadRequest,
+            "Empty PUT request.".to_string(),
```
Comment (acatangiu) [inline api_server/src/request/drive.rs]:
```suggestion
            "Empty PATCH request.".to_string(),
```

File: api_server/src/request/instance_info.rs [hunk 7]
```diff
@@ -0,0 +1,24 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (acatangiu) [inline api_server/src/request/instance_info.rs]:
```suggestion
// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
?

File: api_server/src/request/mmds.rs [hunk 8]
```diff
@@ -0,0 +1,54 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (acatangiu) [inline api_server/src/request/mmds.rs]:
```suggestion
// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api_server/src/request/mmds.rs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+use request::Body;
+use request::Error;
+use request::ParsedRequest;
+
+pub fn parse_get_mmds() -> Result<ParsedRequest, Error> {
```
Comment (acatangiu) [inline api_server/src/request/mmds.rs]: I would drop `instance_info.rs` and `mmds.rs` as separate modules and just treat them in `requests`.
1. They are very small (under 10 lines of production code),
2. Both MMDS and InstanceInfo are ApiServer-local resources.

@firecracker-microvm/compute-capsule what do you say?

File: api_server/swagger/firecracker.yaml [hunk 10]
```diff
@@ -477,6 +477,8 @@ definitions:
           - Uninitialized
           - Starting
           - Running
+          - Halting
+          - Halted
```
Comment (acatangiu) [inline api_server/swagger/firecracker.yaml]: This is merge/rebase damage, these lines should not exist anymore.

File: api_server/src/request/mmds.rs (hunk 9 above)
Comment (alexandruag) [reply api_server/src/request/mmds.rs]: I like the separation approach better, but this is ultimately a highly subjective matter.

Comment (acatangiu) [reply api_server/src/request/mmds.rs]: Ok, keep them as they are.

File: api_server/src/request/drive.rs (hunk 5 above)
Comment (georgepisaltu) [reply api_server/src/request/drive.rs]: I refactored that part and now there is more consistency. Please see if it fits the larger context.

**Pull request merged**